feature/HAB 16 sign up screen

VikkiRy opened the pull request.

VikkiRy committed:
SignUpViewController and its subviews created

VikkiRy committed:
Constraints changed

VikkiRy committed:
Kern values changed

HKurylchykava commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,206 @@
+    }
+    
+    private func configureBackgroundImageView() {
+        backgroundImageView = UIImageView(image: UIImage(named: "mountains"))
```
Check first letter of the names in assets. Images are not displayed

VikkiRy replied:
Images name changed

HKurylchykava commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,206 @@
+    
+    private func configureWelcomLabel() {
+        let label = UILabel()
+        label.translatesAutoresizingMaskIntoConstraints = false
```
I suppose, we shouldn't write label.translatesAutoresizingMaskIntoConstraints = false
with using snapkit

VikkiRy replied:
Removed

VikkiRy committed:
- images changed
- views config and constraints changed

VikkiRy committed:
- translatesAutoresizingMaskIntoConstraints removed
- logoImageView constraints changed

VikkiRy committed:
- lineHeightMultiplyer changed
- logoImageView, welcomeLabel constraints changed

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,212 @@
+НЕ ЗАПОЛНЯЮТ ПУСТОТУ,  ОНИ ЕЕ СОЗДАЮТ!
+НЕ ЗАПОЛНЯЮТ ПУСТОТУ,  ОНИ ЕЕ СОЗДАЮТ!
+ТУТ КАКАЯ-ТО ФРАЗА
+"""
```
Do we plan adding localisation? I suppose we should

VikkiRy replied:
Russian localization added

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,212 @@
+             .paragraphStyle: paragraphStyle
+        ]
+        
+        label.attributedText = NSAttributedString(string: "HABIT HERO", attributes: stringAttributes)
```
Why using attributes instead of setting this values to UILabel?

VikkiRy replied:
I used attributes because I had to change the text kern.
Kern removed (based on your recommendations in pull request #10)

VikkiRy committed:
- Russian localization added;
- iOS deployment target changed to 16.0

VikkiRy committed:
- Inter font name changed;
- Fonts folder moved to supportFiles folder;
- SignUpViewController refactored

VikkiRy committed:
Extra files removed

VikkiRy committed:
Classes for next screen transition added

VikkiRy committed:
- StartButtonConstraints entity added
- startButton constraints refactored

VikkiRy committed:
Contents.json extra files deleted

VikkiRy committed:
Images removed

VikkiRy committed:
Images added

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Colors/Color.swift:
```diff
@@ -0,0 +1,22 @@
+
+extension UIColor {
+    // The CSS color value #CC6699 is written as 0xCC6699 in Swift’s hexadecimal number representation.
+    convenience init(hexNumber: UInt, alpha: CGFloat = 1.0) {
```
For something like that there is a good practice in naming files:
- if your extension contain only one function or init we name file as _ExtendedClass+functionName_ in your example UIColor+hex
- If extension contains several functions we name due to function tasks, like UIColor+convenientInit

VikkiRy replied:
Renamed to UIColor+hex

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/StartButtonConstraints.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+enum StartButtonConstraints: Int {
```
Let's make it more general and rewrite cases as static let.
More general is something like that:
enum LayoutConstants {
static let leadingOffset...
static let trailingOffset...
enum ActionButton {
static let height
static let bottomOffset

VikkiRy replied:
Enum changed

nikGayko commented on BadHabbit/BadHabbit/SupportFiles/ru.lproj/Localizable.strings:
```diff
@@ -0,0 +1,12 @@
+"welcomeLabelText" = "ДОБРО ПОЖАЛОВАТЬ В";
+"phraseLabelText" = "НЕ ЗАПОЛНЯЮТ ПУСТОТУ, ОНИ ЕЕ СОЗДАЮТ! \nНЕ ЗАПОЛНЯЮТ ПУСТОТУ, ОНИ ЕЕ СОЗДАЮТ! \nТУТ КАКАЯ-ТО ФРАЗА";
+"startButtonTitle" = "НАЧАТЬ";
+"logoLabelText" = "ЛОГО";
```
to avoid duplicates in keys add prefix to them, like "signUpScreen.logoLabelText"

VikkiRy replied:
Prefixes added to keys (except for appName)

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenRouter.swift:
```diff
@@ -0,0 +1,29 @@
+
+    func presentAnotherScreen() {
+        let vc = nextScreen.assembleStory()
+        globalCoordinator.presentOnTopVisibleViewController(vc, animated: true, completion: nil)
```
you can use convenient function without params animated and completion

VikkiRy replied:
Did you mean that I can add my own function into GlobalCoordinator class or I can set the default value of the animated parameter as true or something else (in the context of our app architecture)?

nikGayko replied:
I propose to add new function with only one parameter - ViewController

VikkiRy replied:
Function was added

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenRouter.swift:
```diff
@@ -0,0 +1,29 @@
+        self.globalCoordinator = globalCoordinator
+     }
+
+    func presentAnotherScreen() {
```
It's a bad naming. Name function as screen that will be presented + prefix "present"

VikkiRy replied:
Function name changed

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,170 @@
+import SnapKit
+
+protocol SignUpViewProtocol: AnyObject {
+    var startButton: UIButton { get }
```
why it's declared in protocol?

VikkiRy replied:
Because I assumed that I shouldn't declare an empty protocol or wouldn't use the startButtonTapped() function inside as an internal function. Is it correct to add the startButtonTapped() function instead?

nikGayko replied:
There is definitely no need to define variables which aren't used. So it's fine to declare empty protocol

VikkiRy replied:
Variable was removed

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,170 @@
+    private func addAppNameLabel() {
+        view.addSubview(appNameLabel)
+
+        appNameLabel.text = "HABIT HERO".uppercased()
```
Localisation?

VikkiRy replied:
App name localization added

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,170 @@
+        let safeArea = view.safeAreaLayoutGuide
+        
+        backgroundImageView.snp.makeConstraints { make in
+            let heightMultiplier: CGFloat = 445 / 844
```
445 / 844
what this numbers mean?

VikkiRy replied:
It's the ratio of the screen height to background image view height on the mockup. This multiplier is used to calculate the backgroundImageView height.

nikGayko replied:
That's will not work due to different sizes of the screen

VikkiRy replied:
Why? It's just multiplier that's constant and equal 0.52... So, we get a function make.height.equalToSuperview().multipliedBy(0.52...) which should calculate the background image height for different screen sizes.
Otherwise how can we solve this issue?

nikGayko replied:
So here you want to make image height equals to half of screen height?

VikkiRy replied:
In general - yes, but slightly more than half. In this case, can this restriction be applied?

nikGayko replied:
Define multiplier as a constant, not equation

VikkiRy replied:
heightMultiplier was removed, multiplier value was added instead

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,170 @@
+        }
+        
+        phraseLabel.snp.makeConstraints { make in
+            make.bottom.equalTo(startButton.snp.top).inset(-30)
```
let's use only positive numbers in constraints. There is inset and offset, I suppose for you needs offset is fine

VikkiRy replied:
Constraints changed

nikGayko commented:
In general good, besides, tell me if I'm wrong this screen is first one in sign in flow? If so we could open it on init.

VikkiRy committed:
- Color.swift name changed;
- StartButtonConstraints enum refactored;
- Mountains image changed;
- Localizable.strings refactored;
- presentAnotherScreen func renamed;
- Views constraint changed

VikkiRy committed:
startButton constraints changed

VikkiRy committed:
presentSingInAppleScreen function name changed

VikkiRy commented:
> In general good, besides, tell me if I'm wrong this screen is first one in sign in flow? If so we could open it on init.

@nikGayko  Yes, you're right, but only when the user opens the app for the first time. So, as I learned earlier, for example, we can check in SceneDelegate if the user was on SignInAppleScreen and if so, we will change the window.rootViewController. Is it correct?

nikGayko commented:
> > In general good, besides, tell me if I'm wrong this screen is first one in sign in flow? If so we could open it on init.
> 
> @nikGayko Yes, you're right, but only when the user opens the app for the first time. So, as I learned earlier, for example, we can check in SceneDelegate if the user was on SignInAppleScreen and if so, we will change the window.rootViewController. Is it correct?

We will figure out initial screen due to auth status. But so as currently only sing in/sign up flow it's ok to make this screen as root without checks. We will add checks later

VikkiRy committed:
Unused files was removed

VikkiRy committed:
DesignKit folder was changed

VikkiRy committed:
Merge branch 'develop' into feature/HAB-16-sign-up-screen

* develop:
  Added alert assembly, routing mechanism, initial assembly, skeleton for FillingPersonalData

# Conflicts:
#	BadHabbit/BadHabbit.xcodeproj/project.pbxproj

VikkiRy committed:
rootViewController was changed

VikkiRy committed:
Changes in CLANG_ANALYZER_LOCALIZABILITY_NONLOCALIZED was reverted

VikkiRy committed:
project.pbxproj.orig was deleted

VikkiRy committed:
CLANG_ANALYZER_LOCALIZABILITY_NONLOCALIZED was removed

VikkiRy commented:
> > > In general good, besides, tell me if I'm wrong this screen is first one in sign in flow? If so we could open it on init.
> > 
> > 
> > @nikGayko Yes, you're right, but only when the user opens the app for the first time. So, as I learned earlier, for example, we can check in SceneDelegate if the user was on SignInAppleScreen and if so, we will change the window.rootViewController. Is it correct?
> 
> We will figure out initial screen due to auth status. But so as currently only sing in/sign up flow it's ok to make this screen as root without checks. We will add checks later

rootVC was changed to signUpVC

VikkiRy committed:
presentOnTopVisibleViewController() func was added

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Colors/UIColor+hex.swift:
```diff
@@ -0,0 +1,22 @@
+//
+//  Color.swift
```
минорно, имя в комменте осталось прежним

VikkiRy replied:
Name was changed in comment

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/ActionButton.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+enum ActionButton {
```
Имя не соответствует действительности, по сути это только константы layout для ActionButton. Почему не хочешь написать как предложил?

VikkiRy replied:
Хочу, просто предположила, что нужно было выбрать 1 из вариантов. Я поменяю.

VikkiRy replied:
> Let's make it more general and rewrite cases as static let.
> More general is something like that:
> enum LayoutConstants {
> static let leadingOffset...
> static let trailingOffset...
> enum ActionButton {
> static let height
> static let bottomOffset

Я сначала поняла иначе, но теперь предполагаю, что это пример вложенного enum'а.
Если да - тогда не мне не совсем понятно как мы можем применять различные значения к static let leadingOffset.
Я сделала как думаю, что будет правильно, а потом мы по такому же принципу сможем добавлять вложенные enum'ы для других вьюх. Проверь, пожалуйста, верно ли сделано.

nikGayko replied:
Чтобы выровнять все элементы с правой и левой стороны и не было зоопарка отступов(label, картинк, кнопка - все отступают справа с одинаковым оффсетом(если предполагается выравние границы элементы сбоку по границам супервью))

VikkiRy replied:
Исправлено

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenAssembly.swift:
```diff
@@ -0,0 +1,30 @@
+    
+    func assembleStory() -> UIViewController {
+        let router = SignUpScreenRouter(
+            nextScreen: FillingPersonalDataAssembly(alertAssembly: alertAssembly),
```
этот assembly не должен собираться внутри такого же как он assembly. Должен конфигурироваться в StoriesAssembly и передаваться как зависимость

VikkiRy replied:
Dependency was added

VikkiRy committed:
- StartButtonConstraints.swift was changed
- startButton was changed
- Color.swift name was changed

VikkiRy committed:
SignUpScreenAssembly class was changed

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/LayoutConstants.swift:
```diff
@@ -0,0 +1,18 @@
+        static let leadingInset = 16
+        static let trailingInset = 16
+        static let bottomInset = 20
+        static let topOffset = 30
```
This enum's defines global LayoutConstants. In other words constants that used on another screens for such ActionButton. Does topOffset will be reused for another ActionButton in most of the cases?

VikkiRy replied:
No, it doesn't.  topOffset was removed from enum. 
Finally I've got why our enum LayoutConstants look like that :confounded:
Thank you!

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/LayoutConstants.swift:
```diff
@@ -0,0 +1,18 @@
+    enum ActionButton {
+        static let height = 50
+        static let leadingInset = 16
+        static let trailingInset = 16
```
leading and trailing are common not only for ActionButton, but also for labels when leading and trailing constraints equals to superview, so it's LayoutConstants constants

VikkiRy replied:
leadingInset and trailingInset was removed from this enum

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenAssembly.swift:
```diff
@@ -0,0 +1,29 @@
+    }
+    
+    func assembleStory() -> UIViewController {
+        let router = SignUpScreenRouter(nextScreen: nextScreen,
```
Please, don't allign several parameters like that. Start first parameter with new line, other parameters aligned with first one

VikkiRy replied:
Alignment was changed

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenRouter.swift:
```diff
@@ -0,0 +1,29 @@
+
+final class SignUpScreenRouter: SignUpRouterProtocol {
+    
+    let nextScreen: StoryAssembly
```
private?

VikkiRy replied:
Private access was added

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenRouter.swift:
```diff
@@ -0,0 +1,29 @@
+    let nextScreen: StoryAssembly
+    let globalCoordinator: IGlobalCoordinator
+    
+    init(nextScreen: StoryAssembly,
```
first parameter - new line

VikkiRy replied:
Alignment was changed

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,173 @@
+    }
+    
+    private func configureBackground() {
+        view.backgroundColor = UIColor(hexNumber: 0xCFCFCF)
```
If you define color like you will not have adaptive dark theme. Better to define in assets

VikkiRy replied:
darkBackground static constant was aded in UIColor extension

VikkiRy replied:
Colors assets catalog was added

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,173 @@
+        }
+        
+        logoImageView.snp.makeConstraints { make in
+            let multiplayer: CGFloat = 196 / 390
```
It's bad practices to write some numbers which is clear only for you. The called "magic numbers"

https://www.inf.unibz.it/~calvanese/teaching/04-05-ip/lecture-notes/uni04/node15.html#:~:text=A%20magic%20number%20is%20a,difficult%20to%20maintain%20and%20update.

I suppose it's width and height of image. If possible - get this values out of image

VikkiRy replied:
I aded a sizeMultiplier constant and assigned it a closure. Is it possible?

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenRouter.swift:
```diff
@@ -0,0 +1,29 @@
+     }
+
+    func presentSignInAppleScreen() {
+        let vc = nextScreen.assembleStory()
```
we're presenting singIn screen, so rename story assembly to signIn in init parameter and class variable

VikkiRy replied:
Renamed to SignInDataAssembly (as class was named in pr #10)

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenAssembly.swift:
```diff
@@ -0,0 +1,29 @@
+import UIKit
+
+class SignUpScreenAssembly: StoryAssembly {
+    private let nextScreen: StoryAssembly
```
This is signInAssembly

VikkiRy replied:
Renamed to SignInDataAssembly (as class was named in pr #10)

HKurylchykava commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -0,0 +1,173 @@
+        }
+        
+        safeArea.snp.makeConstraints { make in
+            make.centerY.equalTo(logoImageView).offset(35)
```
Is it OK, if we set constraints safeArea equalTo logoImage?

VikkiRy replied:
In
```swift
logoImageView.snp.makeConstraints { make in
            make.centerY.equalTo(safeArea).offset(-35)
}
```
I can't find how can I make this constrains with opposite value like .inset(35).
So for me this solution was the most suitable.
@nikGayko Can we change the safeArea centerY value in this case? Or we have to use for example bottom constraints instead?

nikGayko replied:
Well, nothing bad if we define safeArea equalsTo, or if we define negative constraint value. But there is some more preferable and less preferable solutions, so let's define it in this order.
1. More preferable to {customViewToSetConstraints}.snp.makeConstraints
2. More preferable to define positive value for constraints.

So if we can't suit both points try to keep first and break second.

VikkiRy replied:
Сonstraints was changed

VikkiRy committed:
- enum LayoutConstants was changed
- SignUpScreenRouter nextScreen type was changed
- SignUpViewProtocol was refactored
- opaque was added to logoImageView
- setupViewConstraints() was refactored
- UIColor+darkBackground was added

VikkiRy committed:
- enum LogoImageView was removed
- logoImageView content mode and constraints was changed
- safeArea constraints was removed

VikkiRy committed:
Colors assets catalog was added

VikkiRy committed:
- Colors assets catalog was updated
-UIColor extensions was removed

VikkiRy committed:
- enum BackgroundImage was removed
- backgroundImageView constraints was updated

nikGayko reviewed (approved):
nice job!
